buffer.go

```diff
@@ -0,0 +1,159 @@
+package stats
+
+import (
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+)
+
+// Buffer is the implementation of a measure handler which uses a Serializer to
+// serialize the metric into a memory buffer and write them once the buffer has
+// reached a target size.
+type Buffer struct {
+	// Target size of the memory buffer where metrics are serialized.
+	//
+	// If left to zero, a size of 1024 bytes is used as default (this is low,
+	// you should set this value).
+	BufferSize int
+
+	// The Serializer used to write the measures.
+	//
+	// This field cannot be nil.
+	Serializer Serializer
+
+	buffer unsafe.Pointer
+	join   sync.WaitGroup
+}
+
+// HandleMeasures satisfies the Handler interface.
+func (h *Buffer) HandleMeasures(time time.Time, measures ...Measure) {
+	if len(measures) == 0 {
+		return
+	}
+
+	size := h.BufferSize
+	if size == 0 {
+		size = 1024
+	}
+
+	c := chunkPool.Get().(*chunk)
+	c.b = h.Serializer.AppendMeasures(c.b[:0], time, measures...)
+	l := int32(len(c.b))
+
+	for {
+		b := h.loadBuffer()
+
+		if b == nil {
+			b = bufferPool.Get().(*buffer)
+			b.init(size)
+
+			if !h.compareAndSwapBuffer(nil, b) {
+				bufferPool.Put(b)
+				continue
+			}
+		}
+
+		b.m.RLock()
+		m := int32(len(b.b))
+		n := atomic.AddInt32(&b.n, l)
+
+		if n >= 0 && n <= m {
+			copy(b.b[n-l:n], c.b)
+			b.m.RUnlock()
+			break
+		}
+
+		b.m.RUnlock()
+
+		if n < 0 || (n-l) > m {
+			// Only the goroutine that first overflows the buffer size is in
+			// charge of writing the buffer, the others go back and try to
+			// acquire a new one to copy their measures into.
+			continue
+		}
+
+		if h.compareAndSwapBuffer(b, nil) {
+			b.m.Lock() // wait for read-locks to be released
+			b.m.Unlock()
```

Comment:
The intent of the lock is just to wait for all read locks to be released, so we know that there are no goroutines left in the critical section controlled by the read lock. Once that is done, no more goroutines will ever produce bytes to the buffer (either because they'll see a `nil` value after the compare-and-swap, or because they'll see a full buffer).

After successfully acquiring the write lock the goroutine is the solve owner of the buffer and therefore can release the lock. I agree it's not really intuitive, the intent is to prevent goroutines that would potentially have captured the buffer before it was set to `nil` to block on the flush and therefore move forward and allocate a new buffer on the next loop iteration.